src/v/storage/tests/storage_e2e_test.cc

```diff
@@ -2620,6 +2620,12 @@ FIXTURE_TEST(write_truncate_compact, storage_test_fixture) {
     info("produce_done");
     truncate.get();
     info("truncate_done");
+
+    // Ensure we've cleaned up all our staging segments such that a removal of
```

Comment:
Is it possible to do a more localized unit test of the compaction code that twiddles the generation to force the abort path and validates deletion?  Perhaps not, but it would be nice to have a test that deterministically exercises it.

Reply:
Agreed, it'd be nice to have a better way to reproduce this bug, though I changed approaches so a more targeted test makes a bit less sense for this PR.